File: transports/ruby/test/conformance_test.rb
# frozen_string_literal: true

require_relative "test_helper"

# Verify the Ruby reference library against the checked-in conformance vectors. This
# both guards the encoders/decoders against drift and demonstrates the contract every
# language's reference library follows: reconstruct each vector's envelope from its
# language-neutral `input`, assert encode -> `hex`, and assert decode(`hex`) -> the same
# envelope. Ported from transports/python/tests/test_conformance.py.
module Csilgen
  module Transport
    class RpcVectorsTest < Minitest::Test
      def test_rpc_vectors
        load_vectors("rpc.json").each do |vec|
          name = vec["name"]
          inp = vec["input"]
          expected = vec["hex"]
          case inp["kind"]
          when "request"
            env = RPC::RpcRequest.new(
              service: inp["service"],
              op: inp["op"],
              id: inp["id"],
              payload: unhex(inp["payload_hex"]),
              auth: inp["auth"]
            )
            encoded = env.encode
            assert_equal env, RPC::RpcRequest.decode(encoded), "decode #{name}"
          when "response"
            env = RPC::RpcResponse.new(
              id: inp["id"],
              status: Status.from_code(inp["status"]),
              variant: inp["variant"],
              error: inp["error"],
              payload: unhex(inp["payload_hex"])
            )
            encoded = env.encode
            assert_equal env, RPC::RpcResponse.decode(encoded), "decode #{name}"
          when "push"
            env = RPC::RpcPush.new(
              service: inp["service"],
              event: inp["event"],
              payload: unhex(inp["payload_hex"])
            )
            encoded = env.encode
            assert_equal env, RPC::RpcPush.decode(encoded), "decode #{name}"
          else
            flunk "unknown rpc kind #{inp["kind"]}"
          end
          assert_equal expected, to_hex(encoded), "encode #{name}"
        end
      end
    end

    class EventsVectorsTest < Minitest::Test
      def test_events_vectors
        load_vectors("events.json").each do |vec|
          name = vec["name"]
          inp = vec["input"]
          expected = vec["hex"]
          control = inp["control"]
          if control
            encoded = encode_control(control, inp, name)
            assert_equal expected, to_hex(encoded), "encode #{name}"
            next
          end

          profile = Events::Profile.parse(inp["profile"])
          payload = unhex(inp["payload_hex"])
          env =
            if profile == Events::Profile::VERBOSE
              Events::Event.verbose(inp["service"], inp["event"], payload)
            else
              Events::Event.compact(inp["service_ord"], inp["op_ord"], payload)
            end
          env = env.with_id(inp["id"]) unless inp["id"].nil?
          encoded = env.encode(profile)
          assert_equal expected, to_hex(encoded), "encode #{name}"
          assert_equal env, Events::Event.decode(encoded, profile), "decode #{name}"
        end
      end

      private

      def encode_control(control, inp, name)
        case control
        when "hello"
          Events::Hello.new(
            versions: inp["versions"], profiles: inp["profiles"],
            service: inp["service"], auth: inp["auth"]
          ).encode
        when "hello_ack"
          Events::HelloAck.new(v: inp["v"], profile: inp["profile"], session: inp["session"]).encode
        when "ping"
          Events::Heartbeat.new(nonce: inp["nonce"], at: inp["at"]).encode
        when "close"
          Events::Close.new(status: Status.from_code(inp["status"]), reason: inp["reason"]).encode
        else
          flunk "unknown control #{control} in #{name}"
        end
      end
    end

    class DatagramVectorsTest < Minitest::Test
      def test_datagram_vectors
        load_vectors("datagrams.json").each do |vec|
          name = vec["name"]
          inp = vec["input"]
          expected = vec["hex"]
          case inp["profile"]
          when "cbor-array"
            env = Datagrams::Datagram.new(
              op_ord: inp["op_ord"], seq: inp["seq"], payload: unhex(inp["payload_hex"])
            )
            encoded = env.encode
            assert_equal expected, to_hex(encoded), "encode #{name}"
            assert_equal env, Datagrams::Datagram.decode(encoded), "decode #{name}"
          when "compact-header"
            env = Datagrams::CompactDatagram.new(
              op_ord: inp["op_ord"], seq: inp["seq"],
              epoch: inp["epoch"], body: unhex(inp["body_hex"])
            )
            encoded = env.encode
            assert_equal expected, to_hex(encoded), "encode #{name}"
            assert_equal env, Datagrams::CompactDatagram.decode(encoded), "decode #{name}"
          else
            flunk "unknown datagram profile #{inp["profile"]}"
          end
        end
      end
    end
  end
end

File: transports/dart/test/conformance_test.dart
// Verify the Dart reference library against the checked-in conformance vectors.
//
// This both guards the Dart encoders/decoders against drift and demonstrates the
// contract every reference library follows: reconstruct each vector's envelope
// from its language-neutral `input`, assert encode → `hex`, and assert
// decode(`hex`) → the same envelope. Ported from transports/go/conformance_test.go
// and transports/rust/tests/conformance.rs.

import 'dart:convert';
import 'dart:io';
import 'dart:typed_data';

import 'package:csilgen_transport/csilgen_transport.dart';
import 'package:test/test.dart';

List<Map<String, dynamic>> loadVectors(String name) {
  // `dart test` runs with the package root as the working directory, so the
  // shared vectors sit one level up under conformance/.
  final file = File('../conformance/$name');
  final doc = jsonDecode(file.readAsStringSync()) as Map<String, dynamic>;
  return (doc['vectors'] as List).cast<Map<String, dynamic>>();
}

Map<String, dynamic> input(Map<String, dynamic> vec) =>
    vec['input'] as Map<String, dynamic>;

String? optStr(Map<String, dynamic> m, String k) => m[k] as String?;

int? optInt(Map<String, dynamic> m, String k) => m[k] as int?;

String reqStr(Map<String, dynamic> m, String k) => m[k] as String;

int reqInt(Map<String, dynamic> m, String k) => m[k] as int;

void main() {
  test('RPC vectors', () {
    for (final vec in loadVectors('rpc.json')) {
      final inp = input(vec);
      final name = vec['name'];
      final wantHex = vec['hex'] as String;
      Uint8List actual;

      switch (reqStr(inp, 'kind')) {
        case 'request':
          final req = RpcRequest(reqStr(inp, 'service'), reqStr(inp, 'op'),
              hexToBytes(reqStr(inp, 'payload_hex')));
          req.id = optInt(inp, 'id');
          req.auth = optStr(inp, 'auth');
          actual = req.encode();
          expect(RpcRequest.decode(actual), equals(req),
              reason: '$name decode');
        case 'response':
          final resp = RpcResponse(Status(reqInt(inp, 'status')),
              hexToBytes(reqStr(inp, 'payload_hex')));
          resp.id = optInt(inp, 'id');
          resp.variant = optStr(inp, 'variant');
          resp.error = optStr(inp, 'error');
          actual = resp.encode();
          expect(RpcResponse.decode(actual), equals(resp),
              reason: '$name decode');
        case 'push':
          final push = RpcPush(reqStr(inp, 'service'), reqStr(inp, 'event'),
              hexToBytes(reqStr(inp, 'payload_hex')));
          actual = push.encode();
          expect(RpcPush.decode(actual), equals(push), reason: '$name decode');
        default:
          fail('unknown rpc kind for $name');
      }
      expect(bytesToHex(actual), equals(wantHex), reason: '$name encode');
    }
  });

  test('Events vectors', () {
    for (final vec in loadVectors('events.json')) {
      final inp = input(vec);
      final name = vec['name'];
      final wantHex = vec['hex'] as String;

      // Control payloads are keyed by "control"; application events by "profile".
      final control = optStr(inp, 'control');
      if (control != null) {
        Uint8List bytes;
        switch (control) {
          case 'hello':
            final versions = (inp['versions'] as List).cast<int>();
            final profiles = (inp['profiles'] as List).cast<String>();
            bytes = Hello(versions, profiles,
                    service: optStr(inp, 'service'), auth: optStr(inp, 'auth'))
                .encode();
          case 'hello_ack':
            bytes = HelloAck(reqInt(inp, 'v'), reqStr(inp, 'profile'),
                    session: optStr(inp, 'session'))
                .encode();
          case 'ping':
            bytes =
                Heartbeat(reqInt(inp, 'nonce'), at: optInt(inp, 'at')).encode();
          case 'close':
            bytes = Close(Status(reqInt(inp, 'status')),
                    reason: optStr(inp, 'reason'))
                .encode();
          default:
            fail('unknown control $control');
        }
        expect(bytesToHex(bytes), equals(wantHex), reason: '$name encode');
        continue;
      }

      final profile = parseProfile(reqStr(inp, 'profile'))!;
      final payload = hexToBytes(reqStr(inp, 'payload_hex'));
      Event ev;
      switch (profile) {
        case Profile.verbose:
          ev = Event.verbose(
              optStr(inp, 'service'), reqStr(inp, 'event'), payload);
          ev.id = optInt(inp, 'id');
        case Profile.compact:
          ev = Event.compact(
              reqInt(inp, 'service_ord'), reqInt(inp, 'op_ord'), payload);
          ev.id = optInt(inp, 'id');
      }
      final bytes = ev.encode(profile);
      expect(bytesToHex(bytes), equals(wantHex), reason: '$name encode');
      expect(Event.decode(bytes, profile), equals(ev), reason: '$name decode');
    }
  });

  test('Datagram vectors', () {
    for (final vec in loadVectors('datagrams.json')) {
      final inp = input(vec);
      final name = vec['name'];
      final wantHex = vec['hex'] as String;

      switch (reqStr(inp, 'profile')) {
        case 'cbor-array':
          final d = Datagram(reqInt(inp, 'op_ord'), reqInt(inp, 'seq'),
              hexToBytes(reqStr(inp, 'payload_hex')));
          final bytes = d.encode();
          expect(bytesToHex(bytes), equals(wantHex), reason: '$name encode');
          expect(Datagram.decode(bytes), equals(d), reason: '$name decode');
        case 'compact-header':
          final d = CompactDatagram(reqInt(inp, 'op_ord'), reqInt(inp, 'seq'),
              hexToBytes(reqStr(inp, 'body_hex')));
          final epoch = optInt(inp, 'epoch');
          if (epoch != null) d.withEpoch(epoch);
          final bytes = d.encode();
          expect(bytesToHex(bytes), equals(wantHex), reason: '$name encode');
          expect(CompactDatagram.decode(bytes), equals(d),
              reason: '$name decode');
        default:
          fail('unknown datagram profile for $name');
      }
    }
  });
}
